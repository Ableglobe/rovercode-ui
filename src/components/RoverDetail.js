import React, { Component, Fragment } from 'react';
import {
  Accordion,
  Form,
  Grid,
  Header,
  Icon,
  Loader,
  Message,
  Segment,
  TextArea,
} from 'semantic-ui-react';
import PropTypes from 'prop-types';

import Credential from './Credential';

class RoverDetail extends Component {
  constructor(props) {
    super(props);

    this.state = {
      name: null,
      config: {},
      accordionActive: false,
      configError: false,
      saveError: false,
      saveSuccess: false,
    };
  }

  componentDidMount() {
    const { fetchRover, id } = this.props;

    return fetchRover(id).then(data => this.setState({
      name: data.value.name,
      config: data.value.config,
    }));
  }

  saveRoverSuccess = () => {
    const { fetchRover, id } = this.props;

    this.setState({
      saveSuccess: true,
    });

    fetchRover(id);
  }

  saveRover = () => {
    const { editRover, id } = this.props;
    const { config, configError, name } = this.state;

    if (configError) {
      this.setState({
        saveError: true,
        saveSuccess: false,
      });
      return null;
    }

    this.setState({
      saveError: false,
      saveSuccess: false,
    });

    return editRover(id, {
      name,
      config,
    }).then(this.saveRoverSuccess);
  }

  handleClick = () => {
    const { accordionActive } = this.state;

    this.setState({
      accordionActive: !accordionActive,
    });
  }

  handleConfigChange = (event) => {
    try {
      this.setState({
        config: JSON.parse(event.target.value),
        configError: null,
      });
    } catch (e) {
      this.setState({
        configError: true,
      });
    }
  }

  handleNameChange = (event) => {
    this.setState({
      name: event.target.value,
    });
  }

  render() {
    const { location, rover } = this.props;
    const {
      accordionActive,
      configError,
      saveError,
      saveSuccess,
    } = this.state;

    return (
      <Grid centered divided="vertically" columns={16}>
        <Grid.Row>
          <Header as="h1">
            Rover Settings
          </Header>
        </Grid.Row>
        <Grid.Row>
          <Grid.Column width={4}>
            {
              saveError ? (
                <Grid.Row>
                  <Message negative>
                    <Message.Header>
                      Error found in configuration
                    </Message.Header>
                    <p>
                      Please fix the error before saving.
                    </p>
                  </Message>
                </Grid.Row>
              ) : (null)
            }
            {
              saveSuccess ? (
                <Grid.Row>
                  <Message positive>
                    Rover configuration saved
                  </Message>
                </Grid.Row>
              ) : (null)
            }
            {
              rover && location && location.state && location.state.created ? (
                <Grid.Row>
                  <Message icon info>
                    <Icon name="arrow down" />
                    {`Rover '${rover.name}' has been created. Click the button below to download the credentials.`}
                  </Message>
                </Grid.Row>
              ) : (null)
            }
            {
              rover === null ? (
                <Loader active />
              ) : (
                <Fragment>
                  <Segment raised>
                    <Form key={rover.id} loading={!rover} onSubmit={this.saveRover}>
                      <Form.Input
                        inline
                        label="Name:"
                        defaultValue={rover.name}
                        onChange={this.handleNameChange}
                        required
                      />
                      <Form.Field error={configError}>
                        <Accordion>
                          <Accordion.Title active={accordionActive} onClick={this.handleClick}>
                            <Icon name="dropdown" />
                            Advanced
                          </Accordion.Title>
                          <Accordion.Content active={accordionActive}>
                            <TextArea
                              defaultValue={JSON.stringify(rover.config)}
                              onChange={this.handleConfigChange}
                            />
                          </Accordion.Content>
                        </Accordion>
                      </Form.Field>
                      <Form.Button primary>
                        Save
                      </Form.Button>
                    </Form>
                  </Segment>
                  <Credential rover={rover} />
                </Fragment>
              )
            }
          </Grid.Column>
        </Grid.Row>
      </Grid>
    );
  }
}

RoverDetail.defaultProps = {
  location: null,
  rover: null,
};

RoverDetail.propTypes = {
  editRover: PropTypes.func.isRequired,
  fetchRover: PropTypes.func.isRequired,
  id: PropTypes.number.isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      created: PropTypes.bool.isRequired,
    }),
  }),
  rover: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    config: PropTypes.object.isRequired,
  }),
};

export default RoverDetail;
