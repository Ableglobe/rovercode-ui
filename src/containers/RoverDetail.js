import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies, Cookies } from 'react-cookie';
import { editRover, fetchRover } from '../actions/rover';
import { checkAuthError, authHeader } from '../actions/auth';
import RoverDetail from '../components/RoverDetail';

const mapStateToProps = ({ rover }) => ({ ...rover });
const mapDispatchToProps = (dispatch, { cookies, match }) => ({
  id: parseInt(match.params.id, 10),
  editRover: (id, settings) => dispatch(editRover(id, settings, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
  fetchRover: id => dispatch(fetchRover(id, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
});

const RoverDetailContainer = connect(
  mapStateToProps,
  mapDispatchToProps,
)(RoverDetail);

RoverDetailContainer.propTypes = {
  cookies: PropTypes.instanceOf(Cookies).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default hot(module)(withCookies(RoverDetailContainer));
