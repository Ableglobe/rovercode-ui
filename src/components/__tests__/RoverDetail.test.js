import React from 'react';
import { MemoryRouter } from 'react-router';
import {
  Accordion,
  Dropdown,
  Form,
  Header,
  Loader,
  Message,
  TextArea,
} from 'semantic-ui-react';
import { FormattedMessage } from 'react-intl';
import { mountWithIntl, shallowWithIntl } from 'enzyme-react-intl';
import RoverDetail from '../RoverDetail';

let fetchRover;
let editRover;
let fetchUserList;

describe('The RoverDetail component', () => {
  beforeEach(() => {
    fetchRover = jest.fn(() => Promise.resolve({
      value: {
        name: 'Rover',
        config: {},
      },
    }));
    editRover = jest.fn(() => Promise.resolve({
      value: {
        name: 'Rover',
        config: {},
      },
    }));
    fetchUserList = jest.fn(() => Promise.resolve({
      value: [
        {
          username: 'user1',
        },
        {
          username: 'user2',
        },
      ],
    }));
  });

  test('renders on the page with no errors', () => {
    const rover = {
      id: 1,
      name: 'Sparky',
      config: {
        left_eye_port: 1,
        right_eye_port: 2,
      },
      client_id: '1234',
      client_secret: '5678',
    };
    const wrapper = shallowWithIntl(
      <RoverDetail
        rover={rover}
        fetchRover={fetchRover}
        editRover={editRover}
        fetchUserList={fetchUserList}
        id={1}
      />,
    ).dive();
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches rover on mount', async () => {
    const wrapper = await mountWithIntl(
      <MemoryRouter>
        <RoverDetail
          fetchRover={fetchRover}
          editRover={editRover}
          fetchUserList={fetchUserList}
          id={1}
        />
      </MemoryRouter>,
    );
    expect(fetchRover).toBeCalledWith(1);
    expect(wrapper.find(Loader).exists()).toBe(true);
  });

  test('displays created message', () => {
    const rover = {
      id: 1,
      name: 'Sparky',
      config: {
        left_eye_port: 1,
        right_eye_port: 2,
      },
      client_id: '1234',
      client_secret: '5678',
    };
    const location = {
      state: {
        created: true,
      },
    };
    const wrapper = shallowWithIntl(
      <RoverDetail
        location={location}
        rover={rover}
        fetchRover={fetchRover}
        editRover={editRover}
        fetchUserList={fetchUserList}
        id={1}
      />,
    ).dive();

    expect(wrapper.find(Message).exists()).toBe(true);
    expect(wrapper.find(Message).children().find(FormattedMessage).prop('defaultMessage'))
      .toBe('Rover \'{name}\' has been created. Click the button below to download the credentials.');
    expect(wrapper.find(Message).children().find(FormattedMessage).prop('values')).toEqual({
      name: 'Sparky',
    });
  });

  test('shows the correct information the rover', async () => {
    const rover = {
      id: 1,
      name: 'Sparky',
      config: {
        left_eye_port: 1,
        right_eye_port: 2,
      },
      client_id: '1234',
      client_secret: '5678',
    };
    const wrapper = shallowWithIntl(
      <RoverDetail
        rover={rover}
        fetchRover={fetchRover}
        editRover={editRover}
        fetchUserList={fetchUserList}
        id={1}
      />,
    ).dive();
    await wrapper.instance().componentDidMount();
    wrapper.update();

    expect(wrapper.find(Header).exists()).toBe(true);
    expect(wrapper.find(Loader).exists()).toBe(false);

    expect(wrapper.find(Form.Input).prop('defaultValue')).toBe('Sparky');
    expect(wrapper.find(TextArea).prop('defaultValue')).toEqual(JSON.stringify({
      left_eye_port: 1,
      right_eye_port: 2,
    }));
  });

  test('saves rover settings', async () => {
    const rover = {
      id: 1,
      name: 'Sparky',
      config: {
        left_eye_port: 1,
        right_eye_port: 2,
      },
      client_id: '1234',
      client_secret: '5678',
    };
    const wrapper = shallowWithIntl(
      <RoverDetail
        rover={rover}
        fetchRover={fetchRover}
        editRover={editRover}
        fetchUserList={fetchUserList}
        id={1}
      />,
    ).dive();

    wrapper.find(Form.Input).simulate('change', {
      target: {
        value: 'Marvin',
      },
    });
    wrapper.find(Accordion.Title).simulate('click');
    wrapper.find(TextArea).simulate('change', {
      target: {
        value: JSON.stringify({
          left_eye_port: 3,
          right_eye_port: 4,
        }),
      },
    });
    wrapper.find(Dropdown).simulate('change', {}, {
      value: [
        'user1',
        'user2',
      ],
    });
    wrapper.update();
    await wrapper.instance().saveRover();

    expect(wrapper.state('accordionActive')).toBe(true);
    expect(wrapper.state('saveSuccess')).toBe(true);
    expect(wrapper.state('saveError')).toBe(false);
    expect(wrapper.find(Message).exists()).toBe(true);
    expect(wrapper.find(Message).prop('positive')).toBe(true);
    expect(fetchRover).toHaveBeenCalledWith(1);
    expect(editRover).toHaveBeenCalledWith(1, {
      name: 'Marvin',
      config: {
        left_eye_port: 3,
        right_eye_port: 4,
      },
      shared_users: [
        'user1',
        'user2',
      ],
    });
  });

  test('handles invalid configuration', async () => {
    const rover = {
      id: 1,
      name: 'Sparky',
      config: {
        left_eye_port: 1,
        right_eye_port: 2,
      },
      client_id: '1234',
      client_secret: '5678',
    };
    const wrapper = shallowWithIntl(
      <RoverDetail
        rover={rover}
        fetchRover={fetchRover}
        editRover={editRover}
        fetchUserList={fetchUserList}
        id={1}
      />,
    ).dive();

    expect(wrapper.find(Form.Field).first().prop('error')).toBe(false);

    wrapper.find(TextArea).simulate('change', {
      target: {
        value: 'abcd',
      },
    });

    wrapper.update();

    expect(wrapper.find(Form.Field).first().prop('error')).toBe(true);

    await wrapper.instance().saveRover();
    wrapper.update();

    expect(wrapper.state('saveSuccess')).toBe(false);
    expect(wrapper.state('saveError')).toBe(true);
    expect(wrapper.find(Message).exists()).toBe(true);
    expect(wrapper.find(Message).prop('negative')).toBe(true);
  });
});
