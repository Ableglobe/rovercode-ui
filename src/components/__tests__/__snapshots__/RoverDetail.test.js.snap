// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RoverDetail component renders on the page with no errors 1`] = `
<Grid
  centered={true}
  columns={16}
  divided="vertically"
>
  <GridRow>
    <Header
      as="h1"
    >
      <FormattedMessage
        defaultMessage="Rover Settings"
        id="app.rover_detail.header"
        values={Object {}}
      />
    </Header>
  </GridRow>
  <GridRow>
    <GridColumn
      width={4}
    >
      <GridRow
        style={
          Object {
            "paddingBottom": "10px",
            "paddingTop": "10px",
          }
        }
      >
        <InjectIntl(Credential)
          rover={
            Object {
              "client_id": "1234",
              "client_secret": "5678",
              "config": Object {
                "left_eye_port": 1,
                "right_eye_port": 2,
              },
              "id": 1,
              "name": "Sparky",
            }
          }
        />
      </GridRow>
      <GridRow>
        <Segment
          raised={true}
        >
          <Form
            as="form"
            key="1"
            loading={false}
            onSubmit={[Function]}
          >
            <FormInput
              as={[Function]}
              control={[Function]}
              defaultValue="Sparky"
              inline={true}
              label="Name:"
              onChange={[Function]}
              required={true}
            />
            <FormField
              error={false}
            >
              <Accordion>
                <AccordionTitle
                  active={false}
                  onClick={[Function]}
                >
                  <Icon
                    as="i"
                    name="dropdown"
                  />
                  <FormattedMessage
                    defaultMessage="Advanced"
                    id="app.rover_detail.advanced"
                    values={Object {}}
                  />
                </AccordionTitle>
                <AccordionContent
                  active={false}
                >
                  <TextArea
                    as="textarea"
                    defaultValue="{\\"left_eye_port\\":1,\\"right_eye_port\\":2}"
                    onChange={[Function]}
                    rows={3}
                  />
                </AccordionContent>
              </Accordion>
            </FormField>
            <FormField>
              <FormattedMessage
                defaultMessage="Share with:"
                id="app.rover_detail.shared_user_label"
                values={Object {}}
              />
              <Dropdown
                additionLabel="Add "
                additionPosition="top"
                closeOnBlur={true}
                closeOnEscape={true}
                deburr={false}
                fluid={true}
                icon="dropdown"
                minCharacters={1}
                multiple={true}
                noResultsMessage="No results found."
                onChange={[Function]}
                openOnFocus={true}
                options={Array []}
                renderLabel={[Function]}
                searchInput="text"
                selectOnBlur={true}
                selectOnNavigation={true}
                selection={true}
                value={Array []}
                wrapSelection={true}
              />
            </FormField>
            <FormButton
              as={[Function]}
              control={[Function]}
              primary={true}
            >
              <FormattedMessage
                defaultMessage="Save"
                id="app.rover_detail.save"
                values={Object {}}
              />
            </FormButton>
          </Form>
        </Segment>
      </GridRow>
    </GridColumn>
  </GridRow>
</Grid>
`;
